### ISSUE TITLE
JSON.stringify issue after recent changes

### ISSUE THREAD

**Issue by Ihromant** [konsoletyper/teavm#736]
Hi @konsoletyper 
After recent changes (somewhere between teavm-0.9.0-dev-6 and teavm-0.9.0-dev-12) JS arrays were broken.
I noticed this issue in my serializer. But then I identified that it's not related to it at all.
Here is simple code that reproduces the issue.
```java
@Test
    public void writeArray() {
        JSArray<JSObject> arr = JSArray.create();
        for (int i = 0; i < 3; i++) {
            arr.push(JSNumber.valueOf(Move.values()[(i + 1) % 2].ordinal()));
        }
        assertEquals("[1,0,1]", JSON.stringify(arr));
    }

    private enum Move {
        LEFT, RIGHT
    }
```
Also here is code with broken test in standalone commit. It fixes when we rollback teavm to 0.9.0-dev-6 https://github.com/Ihromant/teavm-io/commit/a2e25f4f922bc4197e3cb832a2aecc2d3ad813cd
I suppose it's related to recent improvement for arrays.

**Comment by konsoletyper** [comment]
Hello. Could not reproduce. What is the actual result of `JSON.stringify`?

**Comment by Ihromant** [comment]
```json
{"$monitor":null,"$id$":0,"$js":{"$monitor":null,"$id$":0,"$js":[0,1,0]}}
```

Strangely, now it fails in test testArrayRecord in teavm-io project, but works for writeArray test.

**Comment by konsoletyper** [comment]
Could you please provide instructions to run this `testArrayRecord`?

**Comment by Ihromant** [comment]
```
mvn clean install
```
in https://github.com/Ihromant/teavm-io project

**Comment by konsoletyper** [comment]
This project is not configured properly - it does not declare TeaVM repository. [Read here](https://teavm.org/docs/intro/preview-builds.html).

**Comment by Ihromant** [comment]
Yes, sorry. I added long ago this repositories to settings.xml, so never tried clean build. Fixed.

**Comment by konsoletyper** [comment]
After researching I found that this issue has nothing to do neither with JSON.serialize, nor with array implementation (which did not change). It's quite sophisticate issue in the edge between metaprogramming and JSO. I don't know the way to fix this easily and unlikely that I'll fix this in the near future. As a workaround you can refuse completely from JSO in metaprogramming code. Write a Java wrapper that hides all private JSO inside, call its public Java method, and everything should be fine.

**Comment by Ihromant** [comment]
Thanks for the quick response.
Then the problem persists. Let's assume that I have any tree-like class (with no cycles, but with transient fields). How to convert it to JSON representation. Example:
```java
class A implements Message {
int b;
int[] c;
transient Map<String, String> cache;
}

A a = new A();
a.b = 1;
a.c = new int[] {2, 3};
```
In need to convert this class somehow to:
```json
{"b": 1, "c": [2,3], "msType": 4328} // msType is added in other place
```
Currently I'm converting it to JSO using metaprogramming. setting msType and then apply JSON.stringify. How to refuse JSO in this code assuming that classes like A can have any structure and any children fields that satisfy the same pattern?

**Comment by konsoletyper** [comment]
You can wrap entire JSO inside some class. You won't refuse from it completely!

**Comment by Ihromant** [comment]
Wrap it before returning from metaprogramming method? Still don't get it.

**Comment by Ihromant** [comment]
Another one question. What changed between 0.9.0-dev-6 and 0.9.0-dev-12? It worked well before.

**Comment by konsoletyper** [comment]
The changes relax JSO. Previously JSObject was not really Object. For example, if you call `hashCode` or `getClass`, you'll get very strange behaviour, perhaps even runtime error. Another issue was that you could not easily pass Object to native JS methods. Series of commit fix these issues. Now TeaVM could transparently wrap, if necessary, JS objects into what *is* actually an Object, namely JSWrapper. What you see is internals of this JSWrapper, because in case of metaprogramming JSO produces excessive wrap instructions. Don't use JSO and Metaprogramming together and all should be fine.

**Comment by Ihromant** [comment]
Ok, let me rephrase the question:
how convert Java object to JSON string representation without JSON.stringify (which now works 50/50) and reflection (which is not supported by teavm)?
Fortunately issue currently is with arrays serialization only (still, need get rid of arrays in all my DTO somehow), but now I'm afraid that one day whole serialization layer of my application will be broken.

**Comment by konsoletyper** [comment]
As I said previously: wrap all JSO inside Java class. Use only this Java class from metaprogramming, avoid direct interaction with wrapped JSO object

**Comment by konsoletyper** [comment]
JSON.stringify works perfectly fine, 100%, not 50/50. You can also refuse from JSO completely by writing you own JSON parser/serializer without re-using one provided by JS

**Comment by Ihromant** [comment]
I found other fix for the issue. https://github.com/Ihromant/teavm-io/commit/8eea07728c402c96e40fc5e1ceea7b9e4684c4b6
Fortunately, it was done by simplifying code.
Many thanks for the assistance.
I don't know whether to close this issue, because generally problem still persists.

**Comment by konsoletyper** [comment]
I hope you won't reproduce this issue anymore after a78eec93d69d62d55b9b1ecf9700fcdc40eafa14

**Comment by Ihromant** [comment]
Great, thanks. Will check it on some occasion.
